File: [0070][Easy][Climbing Stairs]/Climbing_Stairs.py
# leetcode time     cost : 20 ms
# leetcode memory   cost : 11.8 MB 
# Time  Complexity: O(N)
# Space Complexity: O(N)
# solution 2, recursion with memory
class Solution(object):
    def climbStairs(self, n):
        """
        :type n: int
        :rtype: int
        """
        all_ways_cache = {}
        return self.helper(n, all_ways_cache)
    def helper(self,n, all_ways_cache_dict):        
        if n<=2:
            return n
        if n in all_ways_cache_dict.keys():
            return all_ways_cache_dict[n]
        else:
            all_ways_cache_dict[n] = self.helper(n-1, all_ways_cache_dict) + self.helper(n-2, all_ways_cache_dict)
            return all_ways_cache_dict[n] 

def main():
    num = 10          # expect is 89,
    obj = Solution()
    result = obj.climbStairs(num)        
    print("return result is ",result);
    
if __name__ =='__main__':
    main() 

File: [0070][Easy][Climbing Stairs]/Climbing_Stairs.java
// leetcode time     cost : time exceeded.
// leetcode memory   cost : ----
// Time  Complexity: O(2**n)
// Space Complexity: O(n)
// slolution 1, brute force recursion,
public class Solution {
    public int climbStairs(int n) {
        climb_Stairs(0, n);
    }
    public int climb_Stairs(int i, int n) {
        if (i > n) {
            return 0;
        }
        if (i == n) {
            return 1;
        }
        return climb_Stairs(i + 1, n) + climb_Stairs(i + 2, n);
    }
}